### DISCUSSION TITLE
Questions regarding SSR and PocketBase instances. (SvelteKit)

### DISCUSSION THREAD

**Discussion by tomikaka22** [pocketbase/pocketbase#1033] [Q&A] [answered]
I'm a JavaScript newbie and most definitely a SvelteKit newbie. I'm building my first ever app. I have decided to use PocketBase as my database.
I've read the [js-sdk](https://github.com/pocketbase/js-sdk) docs, and at [SSR integration](https://github.com/pocketbase/js-sdk#ssr-integration) it says to "Create a new PocketBase instance for each server-side request".

My question is: **Why not just create a helper file in my "lib" folder that initializes a PocketBase instance, and import that file whenever I want to interact with PB during SSR?**

According to the docs, it's necessary to do the hooks method in order for SSR to work, but surely creating a new instance every request would cause a memory leak? Do instances terminate after the user closes the browser?

I've seen [SupaBase integrated into SvelteKit with the helper file method](https://supabase.com/docs/guides/with-sveltekit#initialize-a-svelte-app), would this not work with PocketBase too? My goal is to have the server access PB only, never the client, so I would only import it in server side code.

Thanks for reading my (probably) dumb/misguided question!

**Comment by ganigeorgiev** [comment, marked as answer]
> Why not just create a helper file in my "lib" folder that initializes a PocketBase instance, and import that file whenever I want to interact with PB during SSR?

You can, but you'll not be able to differentiate the auth state between different client requests (that's why also the pb_auth cookie is used).

---

> According to the docs, it's necessary to do the hooks method in order for SSR to work, but surely creating a new instance every request would cause a memory leak? Do instances terminate after the user closes the browser?

You need to create a separate client instance for each request because node is single threaded by default and the requests are executed in an event loop, meaning that while you are waiting something to get fetched another request may get processed, which could conflicts with your previously instantiated client state. Creating new instances is more or less equivalent to creating a new js object. You don't have to worry about memory leaks since js is garbage collected and the created instance will be deleted after the client reference is no longer used (aka. after returning the response from the node server to the browser).

---

Also please note that if you are starting a new project, it may be better to use the latest pre-release [PocketBase v0.8.0-rc3](https://github.com/pocketbase/pocketbase/releases/tag/v0.8.0-rc3) to avoid manually migrating later due to the breaking changes in v0.8.0. When using PocketBase v0.8.0 you have to use also the [v0.8.0-rc JS SDK](https://github.com/pocketbase/js-sdk/tree/rc) (you can install it with the `next` tag: `npm install pocketbase@next --save`)

**Reply by tomikaka22** [reply]
Thank you for the clear explanation! 
Can you elaborate on the "you'll not be able to differentiate the auth state between different client requests" part?

**Reply by ganigeorgiev** [reply]
@tomikaka22 I'm not sure how to explain it better, but I'll try to give an example.

Let's assume that you have a single PocketBase JS SDK instance in some file and you import this file in all your node server-side actions that interacts with PocketBase.

A single user decides to sends a login request in order to access a private collection, aka. for example you have some action in your SvelteKit code that accept the user credentials and return back a list of records from some private collection.

```js
await client.collection("users").authWithPassword("user1", "123456");
await client.collection("privateCollection").getList(1, 20);
```

After the above call all new requests (from different users too!) will be consider incorrectly as user1 authenticated because in our example the JS SDK client is global and `client.authStore` is already populated with the user1 auth state.

**Reply by tomikaka22** [reply]
Ooh, that makes sense! Thank you for the follow up too, you've been immensely helpful!

**Comment by Amir-Zouerami** [comment]
@ganigeorgiev I have a couple of questions about this. I'm using next.js for context.

1- Do we have to create a new instance in client components (in next.js) as well? or is this a "server components" thing only? If I understand correctly, in client side components, the `authStore` basically reads the `localStorage` from the browser and therefore the value will be different for every user. Meaning the situation you described before (shared memory and multiple users being authenticated after only the first user does) will not happen. correct? If that's the case I should be able to share one pocketbase instance withing my client side components and create new instances for every server side component.

2- In server side components `authStore` is just an object in memory, not some record in the database itself or anything else. correct?

3- Any ideas about storing user auth state in a global state management library like jotai (for client components)? Should I use `authStore.isValid` everytime or can I also store it globally? I need a global state for whether the user is authenticated or not so I can use it everywhere in my app.

**Reply by ganigeorgiev** [reply]
1) It is safe to have a single global client-side SDK instance as there is no risk of mixing the `pb.authStore` state between different users.

2) In general - Yes, PocketBase APIs are fully stateless and we don't store the issued tokens in the database. But note that some JS runtimes (eg. Deno) supports `LocalStorage` and in this case `pb.authStore` will use it.

3) I'm not sure what to recommend here. `pb.authStore.isValid` only loosely checks whether the SDK instance auth store has a token and it's `exp` claim is still valid (but it doesn't guarantee that the token or user associated with it are still existing/valid; this is done when sending the request to the PocketBase server).

**Reply by Amir-Zouerami** [reply]
Thank you. Regarding number 3, what would be the way to guarantee the token? How can I send a request to the pocketbase to make sure?
